docs/README_Step up mechanism.md

```diff
@@ -0,0 +1,56 @@
+# Step up Authentiation
+
+The Step up mechanism is based on the design proposal [1].
+
+What has been implemented (for OIDC):
+* Text mapping in the client configuration to do the text to value translation
+* Putting the achieved level of authentication at the 'acr' claim in both the access token and the id token (either as an integer or an textual representation)
+* Condition Execution to execute the flow (Condition 1, Condition 2, ...)
+* Set Loa on Success Execution (with the option to store the achieved LoA into the user session or not)
+* Extension to the Default Authentication Flow in order to determine if a certain level of authentication is reached
+
+## Auth Request for a Specific Level of Authentication
+To specify a requested level of authentication (LoA) the claims parameter is used in the auth request.
+
+Example Request:
+
+https://{DOMAIN}/auth/realms/{REALMNAME}/protocol/openid-connect/auth?client_id={CLIENT-ID}&redirect_uri={REDIRECT-URI}&scope=openid&response_type=code&response_mode=query&nonce=exg16fxdjcu&claims=%7B%22id_token%22%3A%7B%22acr%22%3A%7B%22essential%22%3Atrue%2C%22values%22%3A%5B%22gold%22%5D%7D%7D%7D
+
+The claims parameter is specified in a JSON representation:
+claims= {
+            "id_token": {
+                "acr": {
+                    "essential": true,
+                    "values": ["silver", "gold"]
+                }
+            }
+        }
+
+Please refer to the openid-spec for further details.
+
+## Flow Configuration
+
+![Flow configuration](Example_Auth_Flow_Step_up.PNG)
+
+## Configure the Text Mapping at the Client
+It is possible to map the acr_values to a textual representation at the client.
+
+If configured, both the access token and the id token will include the acr in a textual representation.
+
+![Client configuration](Client_Config_Step_up.PNG)
+
+
+## Open Points
+
+### No ACR Value Specified in the Request
+The design proposal describes that if no LoA is configured at the client, and in the auth request no LoA (acr_value) is specified, the value -1 shall be used. -1 is a constant that indicates that a full authentication should be performed. 
+This means that Keycloak will always do a full authentication, even if the user is already authenticated to a certain level of authentication, e.g.
```

Comment:
The assumption here is that the client has no LoA configured and also no Loa (ACR) is requested. That means that it is the same case as if the whole feature has not been implemented (current keycloak state). That's why it is also a question of backwards compatibility. Currently keycloak lets you sign in without reauthentication if you have already a valid session (single sign on). This would be broken when the implementation follows the design proposal as it is stated here, because a full reauthentication would be demanded in that case.
I also don't understand your statement that the LoA in the session can't be trusted, because keycloak currently already trusts the session when reauthenticating users. That is the main purpose of the whole single sign on stuff.

Please correct me if I misunderstood you or overlooked something.

Reply:
What I meant is that we should use `0` when doing cookie-based auth (re-authentication). Even though the session was previously granted with an LoA > 0.

For instance, I'm wondering about a situation where the user reached a high level of trust and after a while, he/she is redirected back to re-authenticate. Does not looks correct to keep the same level of trust because you are not attesting the end-user identity but only re-authenticating. In this case, LoA should be lowered to `0` and the user seamlessly authenticated.

Reply:
@pedroigor That's a good point. When re-authentication with a cookie the acr value should be `0`. 

In connection with the cookie issue, we noticed something else:  counter for failed login attempts, which can be activated when using the brute force feature, is be reset by a successful (cookie) authentication. This is very critical from a security perspective when setting up an authentication flow, that has cookie or username-password authentication for level 1 (acr -> 1) and TOTP (acr ->2) for level 2, as the failed login attempts are reset after a successful authentication (in this authentication flow, solely a cookie is sufficient). An attacker would have endless attempts to guess the TOTP, as they can reset the failed login attempts by simply authenticating the user with the cookie (requesting LoA 1). Is this something you are also concerned about, or is it up to the admin configuring the authentication flow? The first solution that came up is to not reset the failed login attempts with a cookie authentication. We created an Jira issue for that: KEYCLOAK-18840. Thanks!

Reply:
A solution could be to return LoA of 0 for re-authentication and to start with the LoA from the user session when doing a step up authentication. This way the user has not to re-enter already provided credentials.

Reply:
@mposolda @pedroigor: We have changed this behavior to the solution presented by @romge, i.e. we always return 0 when re-authenticating via cookie. This also makes us aligned with the OIDC specification.